File: src/kalman_filters/ekf.py
import numpy as np

class EKF:
    """
    Extended Kalman Filter (EKF).
    Suitable for non-linear estimation and navigation.
    """
    def __init__(self, dim_x, dim_z):
        """
        Initialize the EKF.
        dim_x: Dimension of the state vector
        dim_z: Dimension of the measurement vector
        """
        self.dim_x = dim_x
        self.dim_z = dim_z

        self.x = np.zeros(dim_x)
        self.P = np.eye(dim_x)
        self.Q = np.eye(dim_x)
        self.R = np.eye(dim_z)

    def predict(self, fx, F_jac, dt, u=None, Q=None, **kwargs):
        """
        Predict step.
        fx: State transition function f(x, dt, u, **kwargs) -> x_new
        F_jac: Function that returns the Jacobian of f at (x, dt, u) -> F matrix
        dt: Time step
        u: Optional control input
        Q: Optional process noise covariance
        """
        if Q is None: Q = self.Q

        # State Prediction: x = f(x, dt, u)
        self.x = fx(self.x, dt, u, **kwargs)

        # Covariance Prediction: P = FPF' + Q
        F = F_jac(self.x, dt, u, **kwargs)
        self.P = np.dot(np.dot(F, self.P), F.T) + Q

    def update(self, z, hx, H_jac, R=None, **kwargs):
        """
        Update step.
        z: Measurement vector
        hx: Measurement function h(x, **kwargs) -> z_pred
        H_jac: Function that returns the Jacobian of h at x -> H matrix
        R: Optional measurement noise covariance
        """
        if R is None: R = self.R

        # Innovation: y = z - h(x)
        y = z - hx(self.x, **kwargs)

        # Jacobian H
        H = H_jac(self.x, **kwargs)
            
        # Innovation Covariance: S = HPH' + R
        S = np.dot(np.dot(H, self.P), H.T) + R
        
        # Kalman Gain: K = PH'S^-1
        K = np.dot(np.dot(self.P, H.T), np.linalg.inv(S))
        
        # State Correction: x = x + Ky
        self.x = self.x + np.dot(K, y)
        
        # Covariance Correction: P = (I - KH)P(I - KH)' + KRK' (Joseph Form)
        I = np.eye(self.dim_x)
        I_KH = I - np.dot(K, H)
        self.P = np.dot(np.dot(I_KH, self.P), I_KH.T) + np.dot(np.dot(K, R), K.T)

File: cpp/src/ekf.cpp
#include "ekf.hpp"

EKF::EKF(int dim_x, int dim_z) : dim_x(dim_x), dim_z(dim_z) {
    x = Eigen::VectorXd::Zero(dim_x);
    P = Eigen::MatrixXd::Identity(dim_x, dim_x);
    Q = Eigen::MatrixXd::Identity(dim_x, dim_x);
    R = Eigen::MatrixXd::Identity(dim_z, dim_z);
}

void EKF::predict(StateTransitionFunc fx, StateJacobianFunc F_jac, const Eigen::VectorXd& u, const Eigen::MatrixXd& Q_in) {
    // 1. Predict State: x = f(x, u)
    x = fx(x, u);

    // 2. Get Jacobian: F = F_jac(x, u)
    Eigen::MatrixXd F = F_jac(x, u);

    // 3. Predict Covariance: P = FPF' + Q
    const Eigen::MatrixXd& Q_curr = (Q_in.size() > 0) ? Q_in : Q;
    P = F * P * F.transpose() + Q_curr;
}

void EKF::update(const Eigen::VectorXd& z, MeasurementFunc hx, MeasurementJacobianFunc H_jac, const Eigen::MatrixXd& R_in) {
    // 1. Get Measurement Prediction: z_pred = h(x)
    Eigen::VectorXd z_pred = hx(x);

    // 2. Innovation: y = z - z_pred
    Eigen::VectorXd y = z - z_pred;

    // 3. Get Jacobian: H = H_jac(x)
    Eigen::MatrixXd H = H_jac(x);

    // 4. Innovation Covariance: S = HPH' + R
    const Eigen::MatrixXd& R_curr = (R_in.size() > 0) ? R_in : R;
    Eigen::MatrixXd S = H * P * H.transpose() + R_curr;

    // 5. Kalman Gain: K = PH'S^-1
    Eigen::MatrixXd K = P * H.transpose() * S.inverse();

    // 6. State Correction: x = x + Ky
    x = x + K * y;

    // 7. Covariance Correction: P = (I - KH)P(I - KH)' + KRK' (Joseph Form)
    Eigen::MatrixXd I = Eigen::MatrixXd::Identity(dim_x, dim_x);
    Eigen::MatrixXd I_KH = I - K * H;
    P = I_KH * P * I_KH.transpose() + K * R_curr * K.transpose();
}